com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/LaunchRequestHandler.java

```diff
@@ -113,64 +122,153 @@
+                        Arrays.asList(launchArguments.modulePaths),
+                        Arrays.asList(launchArguments.classPaths),
+                        launchArguments.cwd,
+                        envVars);
+                context.setDebugSession(debugSession);
+
+                logger.info("Launching debuggee VM succeeded.");
+
+                ProcessConsole debuggeeConsole = new ProcessConsole(debugSession.process(), "Debuggee", context.getDebuggeeEncoding());
+                debuggeeConsole.onStdout((output) -> {
+                    // When DA receives a new OutputEvent, it just shows that on Debug Console and doesn't affect the DA's dispatching workflow.
+                    // That means the debugger can send OutputEvent to DA at any time.
+                    context.sendEvent(Events.OutputEvent.createStdoutOutput(output));
+                });
+
+                debuggeeConsole.onStderr((err) -> {
+                    context.sendEvent(Events.OutputEvent.createStderrOutput(err));
+                });
+                debuggeeConsole.start();
+
+                return CompletableFuture.completedFuture(response);
+            }
         } catch (IOException | IllegalConnectorArgumentsException | VMStartException e) {
+            logger.log(Level.SEVERE, String.format("Failed to launch debuggee VM. Reason: %s", e.toString()));
             return AdapterUtils.createAsyncErrorResponse(response, ErrorCode.LAUNCH_FAILURE,
                         String.format("Failed to launch debuggee VM. Reason: %s", e.toString()));
         }
-
-        ISourceLookUpProvider sourceProvider = context.getProvider(ISourceLookUpProvider.class);
-        Map<String, Object> options = new HashMap<>();
-        options.put(Constants.DEBUGGEE_ENCODING, context.getDebuggeeEncoding());
-        if (launchArguments.projectName != null) {
-            options.put(Constants.PROJECTNAME, launchArguments.projectName);
-        }
-        sourceProvider.initialize(context.getDebugSession(), options);
-
-        // Send an InitializedEvent to indicate that the debugger is ready to accept configuration requests
-        // (e.g. SetBreakpointsRequest, SetExceptionBreakpointsRequest).
-        context.sendEvent(new Events.InitializedEvent());
-        return CompletableFuture.completedFuture(response);
     }
 
     private static String parseMainClassWithoutModuleName(String mainClass) {
         int index = mainClass.indexOf('/');
         return mainClass.substring(index + 1);
     }
 
+    private String[] constructLaunchCommands(LaunchArguments launchArguments, boolean serverMode, String address) {
+        String slash = System.getProperty("file.separator");
+
+        List<String> launchCmds = new ArrayList<>();
+        launchCmds.add(System.getProperty("java.home") + slash + "bin" + slash + "java");
+        launchCmds.add("-agentlib:jdwp=transport=dt_socket,server=" + (serverMode ? "y" : "n") + ",suspend=y,address=" + address);
```

Comment:
Using String.format for these kind of string manipulation

Reply:
fixed.